### DISCUSSION TITLE
EOM GAJOE differences in results between versions

### DISCUSSION THREAD

**Discussion by I-G-E** [biosaxs-com/atsas-community#1961] [Mixtures and flexible systems]
Hi,

I have a question about differences in outputs between different versions of EOM GAJOE. I have a pool of models generated by MD simulations and the corresponding experimental SAXS data. 

When I use GAJOE version 2.1 from ATSAS 3.2.1 or GAJOE from ATSAS 4.0.1-2, my output Rg and Dmax distributions for both the selected pool and the starting pool are very 'spiky' (ie the distributions are not particularly smooth), often with 0 values in the middle of the distributions. Running these versions of GAJOE requires running FFMAKER to generate the intensities for MD generated models. 

However when I use GAJOE version 2.1 from ATSAS 3.0.4, my output Rg and Dmax distributions for both the selected pool and starting pool are much smoother. This version of GAJOE has the --pool option for using with an externally generated pool of models.

I have used the same settings (to the best of my knowledge) when running both versions of GAJOE.

Please could you explain what is causing such differences in the output between the versions of GAJOE? And is there any way to use the newer version of GAJOE to generate smoother distributions?

![rg_v3](https://github.com/user-attachments/assets/329d0101-fb1e-44ac-a4dd-c769422ad0c3)
![rg_v3](https://github.com/user-attachments/assets/6865fdbb-2829-41c9-bcb8-70e53092500b)

**Comment by franke-biosaxs** [comment]
There weren't many changes in gajoe in between these versions, just the input files. The first thing that comes to mind would be units; inv. Angstrom vs. inv. nanometer. And or the angular range?

Output of FFMAKER is always in inv. Angstrom. Is your data in inv. nanometer, maybe?

**Reply by I-G-E** [reply]
Data is in inverse angstrom so I don't think that would be causing the problem.

**Reply by franke-biosaxs** [reply]
I'd need to look at the models and the data. Could you please either attach the files here if they can be public or send them to me by mail (replace '-' in my username with '@' and add '.com' at the end).

In particular:
 - at least one of the models used as input for ffmaker
 - the command-line for ffmaker
 - the intensity and size files created by ffmaker
 - the command-line for gajoe

**Reply by I-G-E** [reply]
Thank you - I will email you the files. 

ffmaker --output intensities.csv --pool size_statistics.csv *.pdb

For the new version of GAJOE which produces the spiky distributions I used (which was compared with default settings for the old GAJOE):
gajoe --intensity intensities.csv --size size_statistics.csv x.dat --maximum=20 --minimum=5

**Reply by franke-biosaxs** [reply]
Got them. Units are ok. Would have been too easy. 
While gajoe is running on my side: keep in mind that gajoe is a numerical search. Running it N times, will give you N solutions.

Having said that, that's not it. Apparently I "temporarily disabled smoothing" in gajoe (this is what the ChangeLog says) - a function that takes the spiky results and outputs the "nice" stuff by means of a running average. I did not record why I disabled this, though. Probably because of being flabbergasted that someone might ... prettify ... results like this. And forgot to re-enable it.

I'll quote the relevant code below in case you want to apply it yourself. Please note: IMO, there is no reason that there should be any smooth distribution. Math says there isn't. If you run `nnlsjoe` you will get the best non-negative least squares fit across all models to the provided data:
```
        4411  0.10485128878028657          42.9   145.4    42.2   0.123E+06       1  
        4632  0.17611173691596513          41.1   132.6    43.3   0.123E+06       1  
        4846  0.11891212378675355          37.8   124.0    39.9   0.123E+06       1 
        5570  0.16354741801360928          38.7   119.5    43.2   0.123E+06       1  
        6999  0.17227505935174461          44.1   148.9    65.8   0.123E+06       1  
        7428   2.9412374145276116E-002     44.6   146.1    48.8   0.123E+06       1 
        7763  0.23488999900636470          48.4   163.7    63.8   0.123E+06       1  
```
Output is: column in the size csv file, relative weight, Rg, Dmax, CA-CA distance, MW, model ID in the input file, input file (here removed). Note that the chi^2 of `nnlsjoe` is smaller than any you can get from `gajoe`, as `gajoe` does not necessarily find the global minimum.

From this I would guesstimate a "smaller" and a "larger" population contributing to your data. Depending on how you generated the models, you might want to extend the pool with more "large" one and more "small" ones to better sample the overall population, then re-evaluate the fit?

```FORTRAN
  PURE SUBROUTINE SMOOTH_VALUE (values, ind, n, output)
    INTEGER, INTENT(in) :: ind, n
    REAL(DBL), DIMENSION(:), INTENT(in) :: values
    REAL(DBL), INTENT(out) :: output
   INTEGER :: window, i

   output = 0.0d+0
   window = INT(n/2)
   DO i = ind-window, ind+window
     IF ((i .GE. 1) .AND. (i .LE. size(values))) THEN
       output = output + values(i)
     END IF
   END DO
   output = output/n
  END SUBROUTINE SMOOTH_VALUE
```
This function is called once at each position, taking an average around the current `ind`-ex.

**Reply by I-G-E** [reply]
Thank you very much for your detailed response, and also the suggestions for our data. 

In terms of the 'smoothing' code you posted, how would I apply this to my data? For example if I have the output Rg distribution files, can I apply the code directly to this or do I need to re-run GAJOE?

**Reply by franke-biosaxs** [reply]
> In terms of the 'smoothing' code you posted, how would I apply this to my data? For example if I have the output Rg distribution files, can I apply the code directly to this or do I need to re-run GAJOE?

This is Fortran. You'd need to transform it to whatever programming language you prefer. Python would certainly work. Then feed the "spiky" output one-by-one to obtained the smoothed output. Or you use the gajoe results from the previous release. same difference, really.

**Reply by I-G-E** [reply]
Thank you very much for your help.
